crates/holochain_zome_types/src/element.rs

```diff
@@ -235,13 +237,55 @@ impl HashableContent for SignedHeader {
 }
 
 /// The header and the signature that signed it
-#[derive(Clone, Debug, Eq, Serialize, Deserialize)]
-#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
-pub struct SignedHeaderHashed {
-    /// The hashed but unsigned header.
-    header: HeaderHashed,
+pub type SignedHeaderHashed = SignedHashed<Header>;
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(feature = "test_utils", derive(arbitrary::Arbitrary))]
+/// A type that can be hashed and signed equivalently to a [`Header`].
+/// For example it is possible to hash and sign a [`Create`] by
+/// serializing it in a way that produces the same bytes as if
+/// it was serialized as a [`Header::Create`].
+pub struct SignedHashed<T>
+where
+    T: HashableContent<HashType = holo_hash::hash_type::Header>,
+{
+    /// The hash of this header.
+    pub hash: HeaderHash,
+    /// The data that can be serialized, hashed and signed
+    /// as a [`Header`].
+    pub header: T,
     /// The signature of the header.
-    signature: Signature,
+    pub signature: Signature,
+}
+
+impl<T> SignedHashed<T>
+where
+    T: HashableContent<HashType = holo_hash::hash_type::Header>,
+{
+    /// Create a new signed and hashed header by hashing the header.
+    pub fn new(header: T, signature: Signature) -> Self {
+        let (header, hash) = HoloHashed::from_content_sync(header).into_inner();
+        Self {
+            hash,
+            header,
+            signature,
+        }
+    }
+}
+
+impl<T> SignedHashed<T>
+where
+    T: HashableContent<HashType = holo_hash::hash_type::Header>,
+    Header: From<T>,
+{
+    /// Convert a `SignedHeader<T>` to a [`SignedHeaderHashed`]
+    pub fn into_shh(self) -> SignedHeaderHashed {
+        SignedHeaderHashed {
+            hash: self.hash,
+            header: self.header.into(),
+            signature: self.signature,
+        }
+    }
 }
```

Comment:
I don't think this is a great idea, since this assumes that the Info<Header> impl will also preserve the hash of that thing, which is not true in general. If you're going to provide this method, I think you should re-hash the new content to get a new hash. But if you're only doing this in a few places, it may be better to just explicitly create a new SHH from whatever other data (there are no uses of this in this PR).

Reply:
That's the whole point of this type though is that is does have the same hash for the given header.

Reply:
All my comments including this one are based on the idea that since this is written to be fully generic, it should work that way. We might want to use this generically in the future, and some new person coming into the codebase would reasonably expect to be able to pick up this useful tool and have it just work, and not need to know that it's actually only supposed to work for one type and shouldn't be extended.

Reply:
In light of this type being only about Headers, I see more why you did it this way. Something still feels off about it. Imagine somebody in the future writes an `Into<Header>` impl that doesn't preserve the hash. That would technically be valid, since `Into` says nothing about what invariants must be upheld, but it introduces a subtle bug.

I'm not thinking about the state of the current codebase, I'm thinking about how it will be extended in the future.

Reply:
Hmm good point. I actually wonder if this function is needed at all.